python/hyperon/base.py

```diff
@@ -2,24 +2,57 @@
 
 from .atoms import Atom
 
-class GroundingSpace:
+class PySpace:
```

Comment:
I would suggest allow calling methods of the Python space implementations directly without proxying them through C API. Then current `Space` class could be renamed to the `RustSpace` to designate that it proxies the spaces implemented in Rust. `Space` could be new abstract class which has all methods throwing "not implemented" exception. `GroundingSpace` could be a successor of `RustSpace`. And `PySpace` is a successor of `Space`. Please see the code example here: https://github.com/vsbogd/hyperon-experimental/commit/44093d299c5fbf4ba99cf2731aba29a4d9ced086

In such implementation we don't need `get_payload` because user can get data directly from the specific class. `copy` is also not needed. `query` should return `BindingsSet` instead of map in previous implementation. This is not implemented in the commit above and it is why one of the tests fails.  Also methods of the `RustSpace` could be renamed in order to have same names as in Rust and in `Space`.

Reply:
@Necr0x0Der , could you please also take a look at the Python API.

Reply:
Ultimately it is your (and Alexey's) choice for big architectural decisions like this.  However I feel that the system will be more maintainable and have fewer bugs if we channel and centralize functional pathways as much as possible.

By calling through the Core (Rust) interfaces, we will get full interoperability between spaces regardless of which language they are implemented in.  I.e. a Rust client wouldn’t even need to care if a space is implemented in Python or C, assuming the module-loading is done.

Also this gives us the freedom to handle certain functionality at lower levels where appropriate.  For example, if we call through the core, the Python code doesn't need to handle observer registration tracking because it’s done in the base layer.  As we evolve the Space implementation, I believe this will become increasingly important.

Reply:
Am I misunderstanding your proposal?

Reply:
> By calling through the Core (Rust) interfaces, we will get full interoperability between spaces regardless of which language they are implemented in.

This point is true for my suggestion as well. I just would like to simplify the chain of calls when Python Space is called from Python:
- in current code it is: `client -> Space -> hyperonpy -> C API -> hyperonpy -> PySpace`
- in my suggestion it is: `client -> Space`

When call is made from Rust it is identical in both cases (when space is implemented in Python):
`client -> C API -> hyperonpy -> PySpace`

> Also this gives us the freedom to handle certain functionality at lower levels where appropriate. For example, if we call through the core, the Python code doesn't need to handle observer registration tracking because it’s done in the base layer.

I would keep Rust and Python implementations similar here. In Rust we have `Space` trait and notifying observers is implemented by embedding `SpaceCommon` and calling it from methods.  I think Python spaces could be in the same position here.

If we would like to automatically notify observers when add/remove/replace are called we probably should embed custom space implementation into `SpaceCommon` instead (as it is done in this PR). And in this case I still don't see the necessity of calling PySpace methods via `client -> Space -> hyperonpy -> C API -> hyperonpy -> PySpace` chain. We can implement notifications in `SpaceCommon::add` after calling the `PySpace::add` method. If words are confusing I can write a snippet bit later.

Reply:
Basically the idea I tried to express in a last paragraph in Python:
```py
class Space:

    def __init__(self, space_obj):
        self.cspace = hp.space_new_custom(space_obj)

    def __del__(self):
        hp.space_free(self.cspace)

    # other methods

    def add_atom(self, atom):
        self.space_obj.add(atom)
        hp.space_notify_observers(...)

    # all other methods
```

Reply:
I think I see what you have in mind.  With the current code structure, it will be no problem to change things to the way you propose.  However, looking to the future, It means the space management logic needs to remain very thin at the Rust level if Python clients will be allowed to shortcut past it.

What if the benefit of providing the shortcut calling path?  Is it for performance, or some other functionality / design feature you have in mind?

This is a philosophical choice.  If it were me, I would prefer not to allow the shortcut.  My reason is based on a belief that we might want to create a tighter coupling between a top-level host object and the Spaces in the future.  But if spaces remain self-contained and fully owned by the code that instantiates them, then your proposal should be fine.